nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/pgp/PGPControllerService.java

```diff
@@ -0,0 +1,893 @@
+import org.bouncycastle.openpgp.operator.bc.BcPGPDigestCalculatorProvider;
+import org.bouncycastle.openpgp.operator.jcajce.JcaPGPContentSignerBuilder;
+import org.bouncycastle.openpgp.operator.jcajce.JcaPGPContentVerifierBuilderProvider;
+import org.bouncycastle.openpgp.operator.jcajce.JcaPGPDigestCalculatorProviderBuilder;
+import org.bouncycastle.openpgp.operator.jcajce.JcePBEDataDecryptorFactoryBuilder;
+import org.bouncycastle.openpgp.operator.jcajce.JcePBEKeyEncryptionMethodGenerator;
+import org.bouncycastle.openpgp.operator.jcajce.JcePBESecretKeyDecryptorBuilder;
+import org.bouncycastle.openpgp.operator.jcajce.JcePGPDataEncryptorBuilder;
+import org.bouncycastle.openpgp.operator.jcajce.JcePublicKeyDataDecryptorFactoryBuilder;
+import org.bouncycastle.openpgp.operator.jcajce.JcePublicKeyKeyEncryptionMethodGenerator;
+import org.bouncycastle.util.io.Streams;
+
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.charset.StandardCharsets;
+import java.security.Security;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.zip.Deflater;
+
+
+/**
+ * This class defines an implementation of {@link PGPService} suitable for use with the PGP processors in this package.
+ *
+ * In addition to basic Controller Service functionality this class exposes high-level, stream-based cryptographic operations
+ * to clients.  These operations are encrypt, decrypt, sign, and verify.
+ *
+ * This implementation also provides the options builder methods.  These methods build the various options instances that
+ * correspond to the cryptographic operations.  These option instances allow the controller service to combine the properties
+ * of the processors with the properties of the controller, and then use that combination to parameterize the given operation.
+ * Note that this also keeps the keys and pass-phrases out of processor code completely.
+ *
+ * This class also has a set of private methods for reading key material from disk or input streams.  These methods are
+ * not exposed to the processor clients (but are only package-private at the moment).  Note that the key handling methods
+ * support caching.
+ *
+ * The net effect of combining these behaviors is to reduce the significant coupling to the BC/PGP libraries to just
+ * this class.  This was done intentionally to make maintenance and auditing easier.
+ *
+ */
+@CapabilityDescription("Defines cryptographic key material and cryptographic operations for PGP processors.")
+@Tags({"pgp", "gpg", "encryption", "credentials", "provider"})
+public class PGPControllerService extends AbstractControllerService implements PGPService {
```

Comment:
I think we may want to add the `@Restricted()` annotation to this controller service and add a new value to the `org.apache.nifi.components.RequiredPermission` for `ACCESS_ENCRYPTION_KEY` similar to `ACCESS_KEYTAB`.

Reply:
Added the annotation and new permission.  There was a test case for the merging of `ACCESS_KEYTAB`, so I've extended that to include the new `ACCESS_ENCRYPTION_KEY`.

Reply:
Thanks. We will need to note the new permission in the documentation and add an explanation of this new processor family and the permission to the Migration Guidance when this is released.